### ISSUE TITLE
Version 0.19.5 shared build fails on MinGW-w64

### ISSUE THREAD

**Issue by brechtsanders** [actor-framework/actor-framework#1715]
When building version 0.19.5 on Windows under MSYS2 shell with MinGW-w64 GCC 13.2.0 with the following CMake flags
```raw
-DCMAKE_BUILD_TYPE:STRING=Release -DBUILD_SHARED_LIBS:BOOL=ON -DCAF_ENABLE_EXAMPLES:BOOL=OFF -DCAF_ENABLE_TESTING:BOOL=OFF -DCMAKE_CXX_FLAGS:STRING="-fcommon -D_POSIX_THREAD_SAFE_FUNCTIONS"
```
I get the following output which seems to be releated to `dllexport`/`dllimport` issues.:
[ninja_output.txt](https://github.com/actor-framework/actor-framework/files/13878062/ninja_output.txt)

Note that version 0.19.4 was building just fine.

**Comment by Neverlord** [comment]
Thanks for opening an issue! Looks like we're missing an export declaration somewhere at a first glance. We'll look into it.

**Comment by Neverlord** [comment]
I did get linker errors for a DLL build with MSVC as well. I'm not sure if it's the same issue with MinGW. We should also start using DLL builds on CI, I'm already working on that.

**Comment by Neverlord** [comment]
@brechtsanders I've backported the patch to the 0.19 series in the branch `issue/1715-backport`. Would you mind double-checking that it's working for you as well before we cut an 0.19.6 release?

**Comment by brechtsanders** [comment]
Unfortunately this is the result of a shared build of branch `issue/1715-backport`:
```raw
FAILED: libcaf_test.dll libcaf_test.dll.a
C:\Windows\system32\cmd.exe /C "cd . && D:\Prog\winlibs-gcc13.2.0-posix-msvcrt-11.0.1\mingw64\bin\c++.exe -fcommon -D_POSIX_THREAD_SAFE_FUNCTIONS -O3 -DNDEBUG   -shared -o libcaf_test.dll -Wl,--out-implib,libcaf_test.dll.a -Wl,--major-image-version,0,--minor-image-version,19 libcaf_test/CMakeFiles/test_enums.dir/src/test/block_type_strings.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/and_given.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/and_then.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/and_when.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/block.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/context.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/factory.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/fixture/deterministic.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/fixture/flow.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/given.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/nesting_error.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/outline.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/registry.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/reporter.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/requirement_failed.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/runnable.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/runner.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/scenario.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/scope.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/section.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/test.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/then.cpp.obj CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/when.cpp.obj  libcaf_core/libcaf_core.dll.a  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -Wl,--as-needed -lws2_32 -liphlpapi -lucrtbase && cd ."
D:/Prog/winlibs-gcc13.2.0-posix-msvcrt-11.0.1/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/libcaf_test.dir/libcaf_test/caf/test/runner.cpp.obj:runner.cpp:(.text+0x1d50): undefined reference to `__imp__ZN3caf9to_stringB5cxx11ENS_3pecE'
collect2.exe: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```
Also note that before the error there quite a few warnings like this too:
```raw
build_shared/libcaf_test/src/test/block_type_strings.cpp:12:13: warning: 'std::string caf::test::to_string(block_type)' redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
```

**Comment by brechtsanders** [comment]
I can confirm 1418-backport builds fine now

**Comment by Neverlord** [comment]
The patch is included in 0.19.6 (and already fixed in the main branch).
